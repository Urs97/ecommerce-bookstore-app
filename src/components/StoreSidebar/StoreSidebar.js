import React from 'react'
import './StoreSidebar.scss';
import FilterByPrice from './FilterByPrice/FilterByPrice';
import SidebarBookContainer from './SidebarBookContainer/SidebarBookContainer';

function StoreSidebar({ handleFilterByPrice, handleSetCategoryName, bookData }, ref) {

    // Shuffle array and then select first 3 books
    const shuffledBookData = bookData.sort(() => 0.5 - Math.random());
    let selectedBookData = shuffledBookData.slice(0, 3);
    // Cache data and change it only on category change! TODO
    // Make brand logo on homepage link to home

    return (
        <section className="store-sidebar">
            <section className="book-genres">
                <h4 className="sidebar-title">Book Genres</h4>
                <ul>
                    <li><button onClick={() => handleSetCategoryName('science_fiction')}
                        className="btn-to-link">Science Fiction</button></li>
                    <li><button onClick={() => handleSetCategoryName('history')} 
                        className="btn-to-link">History</button></li>
                    <li><button onClick={() => handleSetCategoryName('short_stories')} 
                        className="btn-to-link">Short Stories</button></li>
                    <li><button onClick={() => handleSetCategoryName('finance')} 
                        className="btn-to-link">Finance</button></li>
                    <li><button onClick={() => handleSetCategoryName('biography')} 
                        className="btn-to-link">Biography</button></li>
                    <li><button onClick={() => handleSetCategoryName('programming')} 
                        className="btn-to-link">Programming</button></li>
                </ul>
            </section>
            <FilterByPrice ref={ref} handleFilterByPrice={handleFilterByPrice}/>
            <SidebarBookContainer bookData={selectedBookData} title="Top Rated Books"/>
            <SidebarBookContainer bookData={selectedBookData} title="Recent Books"/>
        </section>
    )
};

const forwardedStoreSidebar= React.forwardRef(StoreSidebar);

export default forwardedStoreSidebar;