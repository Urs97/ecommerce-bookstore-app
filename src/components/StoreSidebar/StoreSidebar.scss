@import '../../styles/shared.scss';

.store-sidebar {
    background-color: #FFFFFF;
    margin-left: 3em;
    padding: 0 1.875em;
    border-radius: $border-radius;

    & .book-genres {

        & ul {
            list-style-position: inside;
            line-height: 1.75;
            padding-top: 0.75em;
        }
    }
}

.store-sidebar > * {
    border-bottom: $sidebar-divider;
    padding: $vertical-sidebar-padding 0;
}