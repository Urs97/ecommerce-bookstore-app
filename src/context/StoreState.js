import React, { useReducer } from "react";

import StoreContext from "../context/StoreContext";
import { useHttp } from "../hooks/useHttp";
import { ACTIONS, storeReducer } from "../reducers/storeReducer";

const StoreState = (props) => {
  const initialState = {
    category: "science_fiction",
    currentPage: 1,
    filterSliderState: false,
    booksPerPage: 9,
    filteredData: null,
    mobileFilterPopup: false,
  };

  const [state, dispatch] = useReducer(storeReducer, initialState);
  const url = `https://openlibrary.org/subjects/${state.category}.json?limit=27&offset=0`;
  const [isLoading, data] = useHttp(url, `store_${state.category}_data`, [
    state.category,
  ]);

  // Pagination functionality
  const indexOfLastBook = state.currentPage * state.booksPerPage;
  const indexOfFirstBook = indexOfLastBook - state.booksPerPage;
  let currentData = [];
  let currentBooks = [];

  if (!isLoading && data) {
    currentData = state.filteredData ? state.filteredData : data.works;
    currentBooks = currentData.slice(indexOfFirstBook, indexOfLastBook);
  }

  const changeCategory = (newCategory) => {
    dispatch({ type: ACTIONS.CHANGE_CATEGORY, payload: newCategory });
  };

  const changeCurrentPage = (newCurrentPage) => {
    dispatch({ type: ACTIONS.CHANGE_CURRENT_PAGE, payload: newCurrentPage });
  };

  const submitFilterValue = (minPrice, maxPrice, data) => {
    dispatch({
      type: ACTIONS.SUBMIT_FILTER_VALUE,
      payload: { minPrice: minPrice, maxPrice: maxPrice, data: data },
    });
  };

  const toggleMobileFilter = (state) => {
    dispatch({ type: ACTIONS.TOGGLE_MOBILE_FILTER, payload: state });
  };

  return (
    <StoreContext.Provider
      value={{
        state: state,
        isLoading: isLoading,
        data: data,
        currentBooks: currentBooks,
        currentData: currentData,
        changeCategory: changeCategory,
        changeCurrentPage: changeCurrentPage,
        submitFilterValue: submitFilterValue,
        toggleMobileFilter: toggleMobileFilter,
      }}
    >
      {props.children}
    </StoreContext.Provider>
  );
};

export default StoreState;
