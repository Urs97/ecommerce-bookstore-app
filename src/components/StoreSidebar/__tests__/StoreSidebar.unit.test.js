import React from 'react';
import {shallow} from 'enzyme';
import StoreSidebar from '../StoreSidebar';
import StoreState from '../../../context/StoreState';

describe('StoreSidebar', () => {
    test('renders StoreSidebar', () => {
        const wrapper = shallow(
            <StoreState>
                <StoreSidebar />
            </StoreState>
        );
        expect(wrapper).toMatchSnapshot();
    });
});