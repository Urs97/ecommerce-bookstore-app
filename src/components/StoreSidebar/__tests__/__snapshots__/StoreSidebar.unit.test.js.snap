// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreSidebar renders StoreSidebar 1`] = `
<ContextProvider
  value={
    Object {
      "changeCategory": [Function],
      "changeCurrentPage": [Function],
      "currentBooks": Array [],
      "currentData": Array [],
      "data": null,
      "isLoading": false,
      "state": Object {
        "booksPerPage": 9,
        "category": "science_fiction",
        "currentPage": 1,
        "filterSliderState": false,
        "filteredData": null,
        "mobileFilterPopup": false,
      },
      "submitFilterValue": [Function],
      "toggleMobileFilter": [Function],
    }
  }
>
  <StoreSidebar />
</ContextProvider>
`;
